remove db checks on every request and use in memory cache (dev only)

kbrock opened the pull request:
## Before

*Development Only*

On every request, we check every miq_product_feature lookup against the database.
This helps us ensure the developer is using a valid feature.

The problem is these features get looked up a lot.
And they often are a recursive lookup when finding the parent feature.

So on the dashboard, there are around 50 lookups with a wall of
text with 215 more => around 265 database lookups.

Good news: most of these lookups are cached, but pages are added to the log
making them hard to read.

```
[----] D, [2022-06-27T18:20:37.786066 #57148:13f24] DEBUG -- :   CACHE  (0.0ms)  SELECT COUNT(*) FROM "miq_product_features" WHERE "miq_product_features"."identifier" = $1  [["identifier", "dashboard_view"]]
```

## After

Since the identifiers are cached in MiqProduceFeature we are instead using the in-memory cache to do these lookups.

250+ lines of noise disappear from the logs.
And potentially a dozen fewer queries are performed

Fryguy commented on app/helpers/application_helper.rb:
```diff
@@ -105,9 +105,9 @@ def role_allows?(**options)
+    if !Rails.env.production? && features.detect { |feature| !MiqProductFeature.feature_exists?(feature) }
       message = "#{__method__} no feature was found with identifier: #{features.inspect}.  Correct the identifier or add it to miq_product_features.yml."
-      identifiers = MiqProductFeature.all.pluck(:identifier)
+      identifiers = MiqProductFeature.features.keys
```
Does the `MiqProductFeature.features` cached get busted when we have new dynamic features (i.e. those weird ones when you create a new tenant and thus they are tenant specific)?  If so, then this is awesome!

kbrock replied:
Yes, they are invalidated [here](https://github.com/ManageIQ/manageiq/blob/master/app/models/tenant.rb#L65) (I was worried about that for a minute

We really need to get rid of those. Product features per tenant was not the best idea we've had.

Fryguy replied:
Agreed - not sure how we undo that at the moment though.

Fryguy reviewed (approved):
LGTM - also want @jrafanie to review.  Also, this is isolated to dev and test, since the production check will short-circuit everything else.

jrafanie commented:
> The problem is these features get looked up a lot. And they often are a recursive lookup when finding the parent feature.
> 
> So on the dashboard, there are around 50 lookups with a wall of text with 215 more => around 265 database lookups.

So, I'm fine with using a cache here but is the problem that we're doing 50 `role_allows?` checks on the dashboard or that we're using the db to lookup the features on each?  Will we be able to detect problems such as 50 privilege checks on one page after this change?

I agree with the change but think we have a feature check problem that is the real problem.

I'm fine with temporary changes but don't want to lose visibility into poorly designed feature checks.

kbrock commented:
@jrafanie Yes, the fact that we introduced a cache says that we have a feature check problem. And for non `admin` users, it is probably much worse.

We have managed to stream line this code a bit and consolidated it to UserRole and Rbac. There are a few nuances that make me uncomfortable changing it, so I've left a few fundamental problems as is.


Again, this fix is for development only. It has bugged me for a few years, but I introduced this so I can read the logs for another PR w/o all this noise.

jrafanie commented:
@miq-bot cross-repo-tests /providers /core

Fryguy commented:
cross-repo is green and @jrafanie approved in word, so merging
